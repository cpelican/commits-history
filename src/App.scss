.commit-history-container {
    display: flex;
    justify-content: center;

    .commit-list-container {
        padding-left: 0;
        display: flex;
        align-items: center;
        flex-direction: column;
    }
}
