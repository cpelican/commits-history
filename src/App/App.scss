.commit-history-container {
    display: flex;
    flex-direction: column;
    align-items: center;

    .commit-list-container {
        padding-left: 0;
        display: flex;
        align-items: center;
        flex-direction: column;
    }

    .loading-container {
        margin-top: 15rem;
    }

    .error-container {
        margin-top: 1rem;
        background-color: rgba(red, 0.2);
        border: 1px solid red;
        border-radius: 4px;
        padding: 0.5rem 1rem;

        .error-message {
            color: red;
        }
    }
}
