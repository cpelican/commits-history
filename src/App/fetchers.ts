import {API_STORAGE_KEY} from '../components/ApiKeySetterItem';
import {ApiCallState, AppState} from '../types';

export const ViewConfig = {
    url: 'https://api.github.com/repos/cpelican/commits-history/commits',
};

export interface SubscribeArgs {
    handleFetchState: (state: AppState) => void;
    handleApiCallState: (apiState: ApiCallState) => void;
}

export class Fetcher {
    public static requestTimeout = 10_000;
    public static isPendingFetchRequest = false;

    private static async waitAndSubscribe(subscribeArgs: SubscribeArgs) {
        await new Promise((resolve) => setTimeout(resolve, Fetcher.requestTimeout));
        if (Fetcher.isPendingFetchRequest) {
            return;
        }
        await Fetcher.subscribe(subscribeArgs);
    }
    // helped by https://javascript.info/long-polling
    public static async subscribe(subscribeArgs: SubscribeArgs) {
        Fetcher.isPendingFetchRequest = true;
        const secretKey = sessionStorage.getItem(API_STORAGE_KEY),
            response = await fetch(ViewConfig.url, {
            method: 'GET',
            headers: {
                Accept: 'application/vnd.github+json',
                Authorization: `Bearer ${secretKey}`,
            },
        });
        Fetcher.isPendingFetchRequest = false;

        if (!response.ok) {
            subscribeArgs.handleFetchState({errorStatus: response.status, items: null});
            subscribeArgs.handleApiCallState('error');
            if (response.status === 502) {
                await Fetcher.waitAndSubscribe(subscribeArgs);
            }
            return;
        }

        const responseData = await response.json();
        if (response.ok && responseData != null) {
            subscribeArgs.handleFetchState({errorStatus: null, items: responseData});
        }
        await Fetcher.waitAndSubscribe(subscribeArgs);
    }
}
