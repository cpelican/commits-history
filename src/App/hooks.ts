import {useEffect, useRef, useState} from 'react';
import {Fetcher} from './fetchers';
import {ApiCallState, AppState} from '../types';

export function usePolling(secretKey: string | null): [AppState, ApiCallState, (apiState: ApiCallState) => void] {
    const [apiCallState, setApiCallState] = useState<ApiCallState>('idle'),
        [state, setState] = useState<AppState>({items: null, errorStatus: null}),
        prevApiCallState = useRef<ApiCallState>();

    useEffect(() => {
        // Avoid eternal loop.
        if (apiCallState === 'loading') {
            return;
        }
        const isRefreshingPage = apiCallState === 'idle' && secretKey != null,
            hasEnteredApiKey = apiCallState === 'polling' && prevApiCallState.current === 'input-change';

        if (isRefreshingPage || hasEnteredApiKey) {
            setApiCallState('loading');
            Fetcher.subscribe({handleFetchState: setState, handleApiCallState: setApiCallState});
        }

        prevApiCallState.current = apiCallState;
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [apiCallState]);

    return [state, apiCallState, setApiCallState];
}
