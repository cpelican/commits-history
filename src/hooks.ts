import {useEffect, useState} from 'react';
import {Fetcher} from './fetchers';

export interface APIItem {
    sha: string;
    commit: {
        author: {
            date: string;
        };
        message: string;
    };
    author: {
        avatar_url: string;
        login: string;
    };
}

export type UseCommitsApiType = [APIItem[] | null, number | null];

export class Hooks {
    public static useCommitsApi = (): UseCommitsApiType => {
        const [errorStatus, setErrorStatus] = useState<number | null>(null),
            [items, setItems] = useState<APIItem[] | null>([]);

        useEffect(() => {
            Fetcher.subscribe(setItems, setErrorStatus);
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, []);

        return [items, errorStatus];
    };
}
