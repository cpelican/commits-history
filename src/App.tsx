import React from 'react';
import './App.scss';
import {CommitItem} from './components/CommitItem';
import {APIItem, Hooks} from './hooks';

function App() {
    const [items, errorStatus] = Hooks.useCommitsApi();

    function renderCommitItem(item: APIItem) {
        const author = {
            name: item.author.login,
            imageUrl: item.author.avatar_url,
        };
        return (
            <CommitItem
                key={item.sha}
                sha={item.sha}
                date={item.commit.author.date}
                message={item.commit.message}
                author={author}
            />
        );
    }

    function renderCommits() {
        if ((items ?? []).length === 0) {
            return null;
        }
        return <ul className='commit-list-container'>{items?.map(renderCommitItem)}</ul>;
    }

    return (
        <div className='commit-history-container'>
            {errorStatus && <p>{`Error with status ${errorStatus}`}</p>}
            {renderCommits()}
        </div>
    );
}

export default App;
