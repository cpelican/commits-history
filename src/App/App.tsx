import React from 'react';
import './App.scss';
import {ApiKeySetterItem, API_STORAGE_KEY} from '../components/ApiKeySetterItem';
import {CommitItem} from '../components/CommitItem';
import {usePolling} from './hooks';
import {APIItem} from '../types';


export function App() {
    const secretKey = sessionStorage.getItem(API_STORAGE_KEY),
        [state, apiCallState, setApiCallState] = usePolling(secretKey),
        hasCommits = (state.items ?? []).length > 0;

    function renderCommitItem(item: APIItem) {
        const author = {
            name: item.author.login,
            imageUrl: item.author.avatar_url,
        };
        return (
            <CommitItem
                key={item.sha}
                sha={item.sha}
                date={item.commit.author.date}
                message={item.commit.message}
                author={author}
            />
        );
    }

    function renderCommits() {
        if (apiCallState !== 'polling') {
            return;
        }

        if(!hasCommits) {
            return <p>No commits so far</p>;
        }

        return <ul className='commit-list-container'>{state?.items?.map(renderCommitItem)}</ul>;
    }

    function renderApiKeySetterItem() {
        if (state.items != null) {
            return;
        }

        if (apiCallState === 'loading') {
            return;
        }

        return <ApiKeySetterItem setAppState={setApiCallState} />;
    }

    function renderError() {
        if (apiCallState !== 'error') {
            return;
        }

        return (
            <div className='error-container'>
                <p className='error-message'>{`Error with status ${state.errorStatus ?? 'unknown'}, please try again`}</p>
            </div>
        );
    }

    function renderLoading() {
        if (apiCallState !== 'loading') {
            return;
        }
        return <p className='loading-container'>Loading</p>;
    }

    return (
        <div className='commit-history-container'>
            {renderApiKeySetterItem()}
            {renderLoading()}
            {renderError()}
            {renderCommits()}
        </div>
    );
}

