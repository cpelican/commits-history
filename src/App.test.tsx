import React from 'react';
import {render} from '@testing-library/react';
import App from './App';
import {Settings} from 'luxon';
import {Hooks, UseCommitsApiType} from './hooks';


describe('App', () => {
    test.each<UseCommitsApiType>([
      [[], false, ''],
      [[], true, ''],
      [[], true, 'End of the world error!'],
      [[], false, 'End of the world error!'],
      [[], false, null],
      [[{
        sha: '1234',
        commit: {
            author: {
                date: '2022-09-14T04:45:56Z'
            },
            message: 'commit message'
        },
        author: {
            avatar_url: 'www.avatar.com',
            login: 'some name'
        }
      }], false, ''],
    ])('renders as expected whith useCommitApi returning api values %s, isLoading %s, error %s', (apiValue, isLoading, error) => {
      jest.spyOn(Hooks, 'useCommitsApi').mockReturnValue([apiValue, isLoading, error]);
      Settings.now = () => new Date(2022, 12, 25).valueOf();
      const {asFragment} = render(<App />);
      expect(asFragment()).toMatchSnapshot();
    });
});

