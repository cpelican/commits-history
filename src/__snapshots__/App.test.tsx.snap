// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders as expected whith useCommitApi returning api values [ [Object] ], isLoading false, error  1`] = `
<DocumentFragment>
  <div
    class="commit-history-container"
  >
    <ul
      class="commit-list-container"
    >
      <li
        class="commit-item"
      >
        <h4>
          commit message
        </h4>
        <div
          class="additional-info-container"
        >
          <img
            alt="avatar for some name"
            class="additional-info-image"
            src="www.avatar.com"
          />
          <div
            class="additional-info-content"
          >
            <span
              class="author-message"
            >
              some name
            </span>
            <span>
              4 months ago
            </span>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`App renders as expected whith useCommitApi returning api values [], isLoading false, error  1`] = `
<DocumentFragment>
  <div
    class="commit-history-container"
  />
</DocumentFragment>
`;

exports[`App renders as expected whith useCommitApi returning api values [], isLoading false, error End of the world error! 1`] = `
<DocumentFragment>
  <div
    class="commit-history-container"
  >
    <p>
      End of the world error!
    </p>
  </div>
</DocumentFragment>
`;

exports[`App renders as expected whith useCommitApi returning api values [], isLoading false, error null 1`] = `
<DocumentFragment>
  <div
    class="commit-history-container"
  />
</DocumentFragment>
`;

exports[`App renders as expected whith useCommitApi returning api values [], isLoading true, error  1`] = `
<DocumentFragment>
  <div
    class="commit-history-container"
  >
    <p />
  </div>
</DocumentFragment>
`;

exports[`App renders as expected whith useCommitApi returning api values [], isLoading true, error End of the world error! 1`] = `
<DocumentFragment>
  <div
    class="commit-history-container"
  >
    <p />
    <p>
      End of the world error!
    </p>
  </div>
</DocumentFragment>
`;
